src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventSink.cs

```diff
@@ -0,0 +1,243 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// Class to register with the RDT and forward RDT events.
+    /// Handles common conditions before sending out notifications.
+    /// </summary>
+    [Export(typeof(RunningDocumentTableEventSink))]
+    internal class RunningDocumentTableEventSink : IVsRunningDocTableEvents3, IDisposable
+    {
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+        private uint _runningDocumentTableEventsCookie;
+
+        public event EventHandler<RunningDocumentTableEventArgs> OnCloseDocument;
+        public event EventHandler<RunningDocumentTableEventArgs> OnRefreshDocumentContext;
+        public event EventHandler<RunningDocumentTableEventArgs> OnReloadDocumentData;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnBeforeOpenDocument;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnInitializedDocument;
+        public event EventHandler<RunningDocumentTableRenamedEventArgs> OnRenameDocument;
+
+        [ImportingConstructor]
+        public RunningDocumentTableEventSink(IThreadingContext threadingContext, IVsEditorAdaptersFactoryService editorAdaptersFactoryService, SVsServiceProvider serviceProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: true);
+            _runningDocumentTable = (IVsRunningDocumentTable4)serviceProvider.GetService(typeof(SVsRunningDocumentTable));
```

Comment:
if you do this, you probably need Assert foreground check.

Reply:
the constructor of ForegroundThreadAffinitizedObject already does the assert since it passes in assertIsForeground = true